Add Initial Account Abstraction Support

adietrichs opened the pull request:
Adds initial account abstraction support to Solidity 0.6.10:

- A new `account` keyword to mark contracts as account abstraction instances
- A new `paygas` builtin Yul function to be used within inline assembly blocks

Example AA contract:

```
// SPDX-License-Identifier: MIT

pragma solidity 0.6.10;
pragma experimental AccountAbstraction;

account contract NumberStorage {
    uint256 _nonce;
    uint256 _number;

    function set_number(uint256 tx_nonce, uint256 gasPrice, bool fail, uint256 new_number) public {
        require(_nonce == tx_nonce);
        _nonce = tx_nonce + 1;
        assembly { paygas(gasPrice) }
        require(!fail);
        _number = new_number;
    }

    function get_nonce() public view returns (uint256) {
        assembly { paygas(0) }
        return _nonce;
    }

    function get_number() public view returns (uint256) {
        assembly { paygas(0) }
        return _number;
    }
    
    constructor() public payable {}
}
```

adietrichs committed:
Add support for the account abstraction prefix

adietrichs committed:
Add support for the "account" keyword for injection of the account abstraction prefix

adietrichs committed:
Add a builtin function "paygas" to Yul to be used in inline assembly blocks

SamWilsn commented on libevmasm/Assembly.cpp:
```diff
@@ -38,6 +38,11 @@ using namespace solidity::evmasm;
 using namespace solidity::langutil;
 using namespace solidity::util;
 
+const uint8_t accountAbstractionPrefix[37] = {0x33, 0x73, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
```
`Bytecode` instead of `Prefix` maybe?

adietrichs replied:
Isn't prefix more precise though? Because it really is only the prefix to the contract, not the full contract bytecode.
Possibly `accountAbstractionPrefixBytecode`, but that might be too verbose.

SamWilsn replied:
The contents of the variable is bytecode, which might not be obvious since it's just a byte array.

adietrichs replied:
I changed it to `aaPrefixBytecode`.

SamWilsn commented on libevmasm/Assembly.cpp:
```diff
@@ -681,6 +686,9 @@ LinkerObject const& Assembly::assemble() const
 			ret.bytecode.resize(ret.bytecode.size() + 20);
 			break;
+		case AccountAbstractionPrefix:
+			ret.bytecode.insert(ret.bytecode.end(), accountAbstractionPrefix, accountAbstractionPrefix+37);
```
Should either hoist `37` into a constant, or maybe `std::size`, though I don't know if that works, or if solidity is C++17.

adietrichs replied:
Agreed, something like that would be preferable. I don't have that much C++ experience, so I initially went with the simplest solution. Will look into it.

adietrichs replied:
Changed both places where I previously used a hardcoded length to `sizeof(aaPrefixBytecode)`

SamWilsn commented on liblangutil/Token.h:
```diff
@@ -143,6 +143,7 @@ namespace solidity::langutil
 	/* Inline Assembly Operators */                                    \
 	T(AssemblyAssign, ":=", 2)                                         \
 	/* Keywords */                                                     \
+	K(Account, "account", 0)                                           \
```
Little bit of bikeshedding here:
 - `owned`
 - `entrypoint`

adietrichs replied:
Agree that `account` is not ideal. I think I don't like `owned` though, seems like it is rather the opposite, no? Like AA contracts are in a sense self-owned, as opposed to EOAs?
Not sure about `entrypoint`, is that confusing with the entrypoint address?

SamWilsn replied:
Hm, how about `sovereign` or `autonomous` then?

SamWilsn replied:
Or even changing it from a `contract` to an `account`?

adietrichs replied:
> Hm, how about `sovereign` or `autonomous` then?

I was also thinking about `sovereign`! I like it, but it is probably too opaque. Not sure about `autonomous`, that sounds like it is connected to DAOs.

adietrichs replied:
> Or even changing it from a `contract` to an `account`?

Just `account` instead of `account contract` is complicated. If I understand it correctly, `contract` right now is one of three possible types: `contract`, `interface`, `library`. It might be that you want to combine AA with several of these, e.g. `account interface`. That would be impossible if `account` instead is its own fourth type...

adietrichs replied:
I think my preference would be to just keep the keyword naming for the initial version, with the understanding that it is temporary. And then ask for suggestions on the discord after we publish our "local testnet MVP".

adietrichs committed:
Add "AccountAbstraction" as an experimental feature.

adietrichs committed:
Only allow the "account" keyword if the experimental "AccountAbstraction" feature is enabled.

adietrichs committed:
Add an empty prerelease.txt to remove the pre-release version tag.

adietrichs commented:
Added some changes related to the way solidity handles its version information.
Now contracts should start with
```
pragma solidity 0.6.10;
pragma experimental AccountAbstraction;
```
Updated the example contract in the PR description.

adietrichs committed:
Rename accountAbstractionPrefix to aaPrefixBytecode and use sizeof() instead of hard-coded bytecode length.

tecywiz121 commented on .gitignore:
```diff
@@ -1,5 +1,5 @@
 commit_hash.txt
-prerelease.txt
+# prerelease.txt
```
What is prerelease.txt?

adietrichs replied:
The presence of `prerelease.txt` signals a release version of solidity. Necessary to be compatible with `pragma solidity 0.6.10`

tecywiz121 commented on libevmasm/Assembly.cpp:
```diff
@@ -681,6 +681,9 @@ LinkerObject const& Assembly::assemble() const
 			ret.bytecode.resize(ret.bytecode.size() + 20);
 			break;
+		case AccountAbstractionPrefix:
+			ret.bytecode.insert(ret.bytecode.end(), aaPrefixBytecode, aaPrefixBytecode+sizeof(aaPrefixBytecode));
```
_technically_ you need to use `sizeof(aaPrefixBytecode) / sizeof(uint8_t)` because C++ is weird.

adietrichs replied:
I added a dedicated `const size_t aaPrefixLength` (as initially proposed by you), I think that is the cleanest solution

adietrichs committed:
Add const aaPrefixLength representing the length of the aaPrefixBytecode array.
